nexus/src/db/datastore.rs

```diff
@@ -1641,109 +1642,153 @@ impl DataStore {
     /*
      * Network interfaces
      */
-
     pub async fn instance_create_network_interface(
         &self,
         interface: IncompleteNetworkInterface,
-    ) -> CreateResult<NetworkInterface> {
+    ) -> Result<NetworkInterface, NetworkInterfaceError> {
         use db::schema::network_interface::dsl;
+        let query = InsertNetworkInterfaceQuery {
+            interface: interface.clone(),
+            now: Utc::now(),
+        };
+        diesel::insert_into(dsl::network_interface)
```

Comment:
So our discussion yesterday made me realize that the saga _action_ needs to be idempotent in that if we've already created _some_ but not all of the interfaces and then crash, and then we start the action again, it should work.  But my (admittedly quick) read of the code is that this will fail with a conflict error because the NIC already exists.  Is that right?

Reply:
Good point. Yes, I think this will fail with a primary key conflict. It seems sufficient to add an `ON CONFLICT DO NOTHING` expression, assuming the rest of the data is OK? But that does suggest that the "dummy preceding action" shouldn't be a no-op, we should allocate the IDs for the NICs there. That way, we get _exactly_ the same record on the later passes through the action, in the case it died partway through.

Does that all sound correct?

Reply:
Dave and I talked about this more in chat. The proposed solution above isn't sufficient.

The example is this. Imagine a provision that requests two NICs for an instance. That saga crashes, not fails, after creating the first NIC but before the second. The saga recovery machinery will pick up at some point, and just retry that one node, without first unwinding the whole saga. We need a way to avoid database conflicts in this case. The primary key is the obvious one, but there are also unique indexes on `(subnet_id, ip)` and `(instance_id, name)`.

We could modify the query to have a clause that ignores conflict on the `id`, the primary key, but the saga still fail when inserting the first NIC, since the `name` is the same. If the user is explicitly requesting an IP address, rather than letting the system allocate one for them, there is also a conflict on the `(subnet_id, ip)` index. Those are critical to the correctness of the system, so we can't just ignore them in the query in general. But we do need a way for the saga to continue with this node during recovery.

The `DataStore::instance_create_network_interface` function will return a useful error, such as `NetworkInterfaceError::IpAddressNotAvailable`. But it's not enough to check for that conflict (or on the `name`) inside the saga, and consider that a successful NIC allocation. For example, a concurrent request could have grabbed either that IP or the name in the meantime.

The way to fix this seems to be in the query, analogous to how the `update_and_check` module handles conditional updates. One rough idea that I'm going to prototype is updating the statement to first look for a record with the same UUID and return that; and if there is no such record, perform the existing query, failing in the existing ways.

Reply:
This should be resolved in 33ef8f4. @davepacheco I'd love another look-see on those bits!